Changelogtemplate

apigram opened the pull request:
This branch should be ready to ship now. I'm unsure about the compilation instructions for Maven, but all other documentation has been updated to reflect the new changes. A test case and some sample template files are also included in this branch which Maven should automatically run on compilation.

Commit:
Initial Maven commit

Commit:
Implemented templating using mustache templates
	Added sample mustache template files that generate a specific changelog format.
	Migrated source to the Maven build system.
	Changelog issues are now displayed sorted by issue type when using the module changelog, otherwise they are displayed as normal (file changelog).
	Fixed and added new Javadoc.
	Added a basic test case for testing the changelog templating.
	NOTE: the JQL string argument  may require use of '_' for spaces and '[' and ']' for quotation marks (due to how Maven handles quotation marks).

Commit:
Added description of new --changelog-file-name flag parameter.

Commit:
Refined template logic
	Modified how version description is displayed to prevent unusual formatting from occuring (due to escaping newlines for HTML parsing)
	Moved Type class into its own file.
	Modified mustache templates to reflect logic changes.
	Changed main procedure to generate both changelogs in one run (rather than having to rerun the program).
	Further enhanced test case to incorporate a full run of the program (both changelogs).
	Added --changelog-file-name parameter to allow users to specify the path and file to save the file-level changelog to.
	Opened a few accessor functions to allow access to values for test case assertions.
	Added sample changelog file.

Commit:
Merge branch 'changelogtemplate' of https://github.com/mleonard87/JIRAChangelogBuilder into changelogtemplate

Commit:
Added more Javadoc.
Changed templates to use a file that is passed from the command line. Argument 6 is the changelog file template, which is also used for the module changelog by default.
Added an optional argument '--module-template' which is used to override the default module template (the template file passed in as a mandatory argument).
Updated test cases to reflect changes.

Commit:
Updated readme to reflect changes to build system and command-line arguments.

Commit:
Added more detail to compilation instructions.

switchtrue commented on ChangelogTemplate/changelog.txt:
```diff
@@ -0,0 +1,22 @@
+=================================================================
```
Pretty certain this file shouldnt be included.

switchtrue commented on ChangelogTemplate/file.mustache:
```diff
@@ -0,0 +1,11 @@
+=================================================================
+{{dateText}} {{name}}{{{description}}}
+-----------------------------------------------------------------
+{{#issueList}}
```
Could we please normalise names here? You have "versions" but "issueList". It would be nice to have "versions", "issues", "issueTypes", etc.

switchtrue commented on ChangelogTemplate/src/main/java/net/foxopen/jira/changelog/Changelog.java:
```diff
@@ -0,0 +1,119 @@
+    System.out.println("<JIRA_project_name>: The name of the project in JIRA.");
+    System.out.println("<version>: The name of the version this changelog is for.");
+    System.out.println("<flags> (optional): One or more of the following flags:");
+    // TODO: If this JQL causes no issues to be returned, it causes a hard error. Handle this more nicely.
```
Have you done this yet? Can we get that fixed please?

switchtrue commented on ChangelogTemplate/src/main/java/net/foxopen/jira/changelog/Changelog.java:
```diff
@@ -0,0 +1,119 @@
+	/**
+	 * Show usage of the application.
+	 */
+  public static void showUsage()
```
Do you need to update this to include the parameters for the template paths?

switchtrue commented on ChangelogTemplate/src/main/java/net/foxopen/jira/changelog/Changelog.java:
```diff
@@ -0,0 +1,119 @@
+    final String jiraPassword    = args[currentArgument++];
+    final String jiraProjectKey  = args[currentArgument++];
+    final String versionName     = args[currentArgument++];
+		final String fileTemplate		 = args[currentArgument++];
```
So I want to pass the templates in as a CSV list of paths rather than two arguments. So usage would be:

http://company.myjira.com someuser p@ssw0rd! CLB CLB-0.1.01 "/usr/share/templates/file1.txt,/usr/share/tempaltes/file2.xml" <flags>

This will allow us to define as many templates as we like. If none are found in the arguments error.

switchtrue commented on ChangelogTemplate/src/main/java/net/foxopen/jira/changelog/VersionInfoCache.java:
```diff
@@ -0,0 +1,95 @@
+package net.foxopen.jira.changelog;
```
Why are all these files coming up as added? They've always existed?!?

apigram replied:
They're appearing because the project is using a different directory to the original. I'll try to fix this up.

switchtrue commented on ChangelogTemplate/src/test/java/net/foxopen/jira/changelog/ChangelogTemplateTest.java:
```diff
@@ -0,0 +1,108 @@
+/*
```
Nice!

switchtrue commented on README.md:
```diff
@@ -24,7 +23,10 @@ Where the arguments are used as follows:
   *  `<version>`: The name of the version this changelog is for.
+  *  `<file_template>`: The path to a file that will be used as the changelog template.
   *  `<flags>` (optional): One or more of the following flags:
     * '--jql "some arbitrary JQL": Append the given JQL to the issue filter. eg status = "Ready for Build"'
```
this line is broken, its missing the `s around --jql. You can preview markdown here: http://daringfireball.net/projects/markdown/dingus

Commit:
Moved files to make diffs consistent with master branch.

Commit:
Restructured argument handling to use a CSV list for a collection of template paths rather than one.
One changelog per template is now generated, each in their own file (currently with a number identifier).
Updated test case to reflect changes.

Commit:
Moved code and test files around directory so as to preserve the structure of the original Changelog Builder code.
Rewrote the maven POM to reflect the new code and test directories.
Updated test cases to reflect changes.

Commit:
Added rudimentary file check to prevent non-existent temp;late files from being used in the system by attempting to open the file. Non-existent template files are ignored.

Commit:
Fixed file naming in test case noit being consistent with program arguments. If the test case fails, Maven will fail the compilation so this fix is necessary.

Commit:
Changed layout of template files to prevent escaping of special HTML characters in the name and description of an issue (<, >, ', ", etc.).
Changed artifact (jar) name and version number.
Added build 1.04.00.
